File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        int n=nums.size();
        std::map<int ,int>mp;
        int l=0;
        int ans=0;
        for(int r=0;r<n;r++){
            mp[nums[r]]++;
            while(mp[nums[r]]>k){
                mp[nums[l]]--;
                l++;
            }
            ans=std::max(ans, r-l+1);
        }
        return ans;
    }
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {

        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();

        int l = 0;
        int ans = 0;

        for (int r = 0; r < n; r++) {

            mp.put(nums[r], mp.getOrDefault(nums[r], 0) + 1);

            while (mp.get(nums[r]) > k) {
                mp.put(nums[l], mp.get(nums[l]) - 1);
                l++;
            }

            ans = Math.max(ans, r - l + 1);
        }

        return ans;
    }
}